Fix secondary index feature with tablets

nyh opened the pull request:
Before this series, materialized views already work correctly on keyspaces with tablets, but secondary indexes do not. The goal of these series is make CQL secondary indexes fully supported on tablets:

1. First we need to make CREATE INDEX work with tablets (it didn't before this series). Fixes #16396.
2. Then we need to keep the promise that our documentation makes - that **local** secondary index should be synchronously updated - Fixes #16371.

As you can see in the patches below, and as was expected already in the design phase, the code changes needed to make indexes support tablets were minimal. But writing reliable tests for these issues was the biggest effort that went into this series.

nyh commented:
Note to reviewers: Only the last three patches are new in this PR, the previous ones are from previous PRs which haven't got in yet. Please review only the last three patches here.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5283/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6692/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5283/)
#### Failed Tests (36/23520):
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5283/testReport/junit/boost.secondary_index_test.test_local_and_global_secondary_index.3.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5283/testReport/junit/boost.secondary_index_test.test_local_and_global_secondary_index.3.xunit/xml/Tests___Unit_Tests____empty__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5283/testReport/junit/boost.secondary_index_test.test_local_index_case_sensitive.2.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5283/testReport/junit/boost.secondary_index_test.test_local_index_case_sensitive.2.xunit/xml/Tests___Unit_Tests____empty__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5283/testReport/junit/boost.secondary_index_test.test_local_index_multi_pk_columns.2.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5283/testReport/junit/boost.secondary_index_test.test_local_index_multi_pk_columns.2.xunit/xml/Tests___Unit_Tests____empty__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5283/testReport/junit/boost.secondary_index_test.test_local_index_paging.2.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5283/testReport/junit/boost.secondary_index_test.test_local_index_paging.2.xunit/xml/Tests___Unit_Tests____empty__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5283/testReport/junit/boost.secondary_index_test.test_local_index_unorthodox_name.3.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5283/testReport/junit/boost.secondary_index_test.test_local_index_unorthodox_name.3.xunit/xml/Tests___Unit_Tests____empty__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
#### Build Details:
- Duration: 2 hr 20 min
- Builder: spider8.cloudius-systems.com

tgrabiec commented on db/schema_tables.cc:
```diff
@@ -2691,14 +2691,32 @@ static void make_update_indices_mutations(
         auto& cf = db.find_column_family(new_table);
         auto view = cf.get_index_manager().create_view_for_index(index, new_token_column_computation);
         auto view_mutations = make_view_mutations(view, timestamp, true);
         view_mutations.copy_to(mutations);
```
Maybe worth extracting this to a function/lambda to avoid duplication?

nyh replied:
Maybe, I'll take a look how to do that in the least ugly way. Although the duplicates are just a few lines from each other so it's hard to make a mistake.

nyh replied:
I'll do it with a lambda, and send a new version of this PR in a few minutes.

tgrabiec commented on test/topology_experimental_raft/test_mv_tablets.py:
```diff
@@ -7,15 +7,74 @@
+    # listed in different system tables so we may need to search both:
+    matches = list(await manager.cql.run_async(f"select id from system_schema.tables where keyspace_name = '{keyspace_name}' and table_name = '{table_or_view_name}'"))
+    if not matches:
+        matches = list(await manager.cql.run_async(f"select id from system_schema.views where keyspace_name = '{keyspace_name}' and view_name = '{table_or_view_name}'"))
```
You could extend ManagerClient.get_table_id().

nyh replied:
That's exactly why I don't like putting convenience functions in "global" places - unless very well thought out, when the next user wants to use them they never do the right thing, and then they start to get 17 different options and features that people needed in 17 different tests. (again, this is my bad experience from dtest speaking..).

In this case, as you guessed I tried to use get_table_id() but it wasn't good enough because it didn't search for views. You're right I could have improved get_table_id() but the question is - why. Why should the get_table_id() functionality exist in the "manager"? Why is it the job of the cluster manager to provide 100 different convenience functions that perform specific CQL requests?

nyh replied:
By the way, I just realized that this function didn't need to use the concept of "table_id" at all this complicated way to fetch it. If could just look for the given `table_name` and `keyspace_name` (but need to use `ALLOW FILTERING`, this becomes a whole-table scan of the tablets table). This would find both tables and view names.

(Another by the way: If you do use table_id like you do today, you don't need to also filter the keyspace when reading the tablets table - a table_id can't possibly belong to other keyspaces).

nyh replied:
I decided to stop using table_id at all (so in particular I won't need get_table_id()). I'll send that change as a followup patch which will also improve the Alternator LSI tests (to use 2 nodes instead of 10).

tgrabiec replied:
> In this case, as you guessed I tried to use get_table_id() but it wasn't good enough because it didn't search for views. You're right I could have improved get_table_id() but the question is - why. Why should the get_table_id() functionality exist in the "manager"? Why is it the job of the cluster manager to provide 100 different convenience functions that perform specific CQL requests?

Why extend get_table_id() - because new logic fits into its semantics. 

Why have get_table_id() - It's good to have common vocabulary for working with scylla, so that tests are easier to write, read and maintain. When I see get_table_id() I immediately know what is intended. When writing a test, I don't have to figure out the implementation on my own (column names?). It's also more robust against changes of the system.  For example, you added another implementation of get_tablet_replicas(). If we change the layout of system.tablets, we now have to update it in two places rather than one.

Why manager - it doesn't have to live in the manager, it could live somewhere else. It's common to all tests which work with scylla.

> (Another by the way: If you do use table_id like you do today, you don't need to also filter the keyspace when reading the tablets table - a table_id can't possibly belong to other keyspaces).

Keyspace name is part of primary key, so it had to be specififed.

nyh replied:
I changed the test - https://github.com/scylladb/scylladb/pull/16440/commits/a24c16333865937b68bc2409e34561762711d01f - to completely avoid a table_id

Basically you can just look up (with ALLOW FILTERING - this is a full table scan) given table_name and keyspace_name - no need to look up the table_id. Maybe the table_id is useful when there might be thousands of tables in the system but this doesn't happen in our test.

kostja replied:
I wonder if we want to follow up with the changelog note re this change. If you agree it's the case, please update @tzach

nyh commented:
Pushed a new version with:
1. Rebased on new master, so now this PR only has the right patches - no unrelated stuff from other PRs.
2. Fixed review comments - reduced code duplication and dropped the function get_table_id (and it use) in a test.

Guys, please review.

nyh commented:
Oops, the first patch is still garbage, I'll remove it and push again, sorry about the spam.

scylladb-promoter commented:
#### :red_circle: CI State: ABORTED

#### Build Details:
- Duration: 28 min
- Builder: spider1.cloudius-systems.com

nyh commented:
> #### 🔴 CI State: ABORTED

I don't know what this abort means. I also see that unrelated tests failed before the abort. I'll restart the CI and hope for the best.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5341/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6756/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5341/)
#### Failed Tests (33/23551):
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5341/testReport/junit/boost.secondary_index_test.test_local_and_global_secondary_index.1.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5341/testReport/junit/boost.secondary_index_test.test_local_and_global_secondary_index.2.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5341/testReport/junit/boost.secondary_index_test.test_local_and_global_secondary_index.3.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5341/testReport/junit/boost.secondary_index_test.test_local_index_case_sensitive.1.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5341/testReport/junit/boost.secondary_index_test.test_local_index_case_sensitive.1.xunit/xml/Tests___Unit_Tests____empty__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5341/testReport/junit/boost.secondary_index_test.test_local_index_case_sensitive.2.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5341/testReport/junit/boost.secondary_index_test.test_local_index_case_sensitive.3.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5341/testReport/junit/boost.secondary_index_test.test_local_index_case_sensitive.3.xunit/xml/Tests___Unit_Tests____empty__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5341/testReport/junit/boost.secondary_index_test.test_local_index_multi_pk_columns.1.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5341/testReport/junit/boost.secondary_index_test.test_local_index_multi_pk_columns.1.xunit/xml/Tests___Unit_Tests____empty__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
#### Build Details:
- Duration: 2 hr 13 min
- Builder: spider8.cloudius-systems.com

nyh commented:
@benipeled or @kostja or somebody, do you know what these "empty" xml files mean?

Anyway, the fact those are failing secondary index unit tests - and my patch modified secondary index, suggests something's wrong in my patch. I'll take a look at it.

kostja commented:
An empty xml file means an exception or a failure to generate the file contents.

nyh commented:
> An empty xml file means an exception or a failure to generate the file contents.

Exception or failure by whom and why?

Anyway, I can reproduce this failure locally, for example:

```
build/dev/test/boost/secondary_index_test --run_test=test_local_index_multi_pk_columns -- -c1 -m1G
```

crashes with a segfault. I'll debug it.

nyh commented:
Pushed a new version with a new third patch, fixing cql_test_env to add the "tags" extension which we forgot - and will now be needed for the LSI tests because LSIs will get the synchronous_updates tag. This should fix the failure Boost tests that we saw in the previous version of this PR.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5349/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6766/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5349/)
#### Build Details:
- Duration: 2 hr 25 min
- Builder: spider5.cloudius-systems.com

nyh commented:
Good - after I fixed the cql_test_env bug, the CI now passes. Please review - and we'll have tablets fully working (almost, sort of) for materialized views, secondary index and also Alternator :-)

kostja commented on db/view/view.cc:
```diff
@@ -1609,10 +1609,13 @@ static future<> apply_to_remote_endpoints(service::storage_proxy& proxy, locator
+    // The "delay_before_remote_view_update" injection point can be
+    // used to add a short delay (currently 0.5 seconds) before a base
+    // replica sends its update to the remove view replica.
+    co_await utils::get_local_injector().inject("delay_before_remote_view_update", 500ms);
```
remove -> remote

nyh replied:
I'll fix this in my followup PR, thanks.

tgrabiec commented on test/lib/cql_test_env.cc:
```diff
@@ -109,6 +109,7 @@ cql_test_config::cql_test_config(shared_ptr<db::config> cfg)
 
     db_config->add_cdc_extension();
     db_config->add_per_partition_rate_limit_extension();
+    db_config->add_tags_extension();
```
Why can't we add all extensions here, and call the same add_all_extensions() from main?

Preexisting problem, so not a blocker.
